nimbus/src/enrollment.rs

```diff
@@ -85,149 +441,232 @@ pub fn get_enrollments(db: &Database) -> Result<Vec<EnrolledExperiment>> {
+        all_slugs.extend(existing_experiments.keys());
+        all_slugs.extend(updated_experiments.keys());
+        all_slugs.extend(existing_enrollments.keys());
+
+        for slug in all_slugs {
+            let updated_enrollment = self.evolve_enrollment(
+                is_user_participating,
+                existing_experiments.get(slug).copied(),
+                updated_experiments.get(slug).copied(),
+                existing_enrollments.get(slug).copied(),
+                &mut enrollment_events,
+            )?;
+            if let Some(enrollment) = updated_enrollment {
+                updated_enrollments.push(enrollment);
+            }
+        }
+
+        Ok((updated_enrollments, enrollment_events))
+    }
+
+    /// Evolve a single enrollment using the previous and current state of an experiment.
+    fn evolve_enrollment(
+        &self,
+        is_user_participating: bool,
+        existing_experiment: Option<&Experiment>,
+        updated_experiment: Option<&Experiment>,
+        existing_enrollment: Option<&ExperimentEnrollment>,
+        enrollment_events: &mut Vec<EnrollmentChangeEvent>, // out param containing the events we'd like to emit to glean.
+    ) -> Result<Option<ExperimentEnrollment>> {
+        Ok(
+            match (existing_experiment, updated_experiment, existing_enrollment) {
+                // New experiment.
+                (None, Some(experiment), None) => Some(ExperimentEnrollment::from_new_experiment(
+                    is_user_participating,
+                    self.nimbus_id,
+                    self.available_randomization_units,
+                    self.app_context,
+                    experiment,
+                    enrollment_events,
+                )?),
+                // Experiment deleted remotely.
+                (Some(_), None, Some(enrollment)) => {
+                    enrollment.on_experiment_ended(enrollment_events)
+                }
+                // Known experiment.
+                (Some(_), Some(experiment), Some(enrollment)) => {
+                    Some(enrollment.on_experiment_updated(
+                        is_user_participating,
+                        self.nimbus_id,
+                        self.available_randomization_units,
+                        self.app_context,
+                        experiment,
+                        enrollment_events,
+                    )?)
+                }
+                (None, None, Some(enrollment)) => enrollment.maybe_garbage_collect(),
+                (None, Some(_), Some(_)) => {
+                    return Err(Error::InternalError(
+                        "New experiment but enrollment already exists.",
+                    ))
+                }
+                (Some(_), None, None) | (Some(_), Some(_), None) => {
+                    return Err(Error::InternalError(
+                        "Every experiment in the DB has an associated enrollment record.",
```

Comment:
micro-nit: when errors occur, it's easier to parse them if they're phrased in terms of what was wrong rather than what was supposed to be right, which in this case would be "Experiment in the db did not have an associated enrollment record")